subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,740 @@
+
+	atomic_set_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
+	ddata->protocol = HID_PROTOCOL_REPORT;
+	if (ops->iface_ready) {
+		ops->iface_ready(dev, true);
+	}
+
+	if (desc->out_ep.bLength != 0U) {
+		k_work_submit(&ddata->output_work);
+	}
+
+	LOG_DBG("Configuration enabled");
+}
+
+static void usbd_hid_disable(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *const ops = ddata->ops;
+
+	atomic_clear_bit(&ddata->state, HID_DEV_CLASS_ENABLED);
+	if (ops->iface_ready) {
+		ops->iface_ready(dev, false);
+	}
+
+	LOG_DBG("Configuration disabled");
+}
+
+static void usbd_hid_suspended(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+
+	LOG_DBG("Configuration suspended, device %s", dev->name);
+}
+
+static void usbd_hid_resumed(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+
+	LOG_DBG("Configuration resumed, device %s", dev->name);
+}
+
+static void *usbd_hid_get_desc(struct usbd_class_data *const c_data,
+			       const enum usbd_speed speed)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+
+	if (speed == USBD_SPEED_HS) {
+		return ddata->hs_desc;
+	}
+
+	return ddata->fs_desc;
+}
+
+static int usbd_hid_init(struct usbd_class_data *const c_data)
+{
+	LOG_DBG("HID class %s init", c_data->name);
+
+	return 0;
+}
+
+static void usbd_hid_shutdown(struct usbd_class_data *const c_data)
+{
+	LOG_DBG("HID class %s shutdown", c_data->name);
+}
+
+static struct net_buf *hid_buf_alloc_ext(struct hid_device_data *const ddata,
+					 const uint16_t size, void *const data,
+					 const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc_with_data(ddata->pool_in, data, size, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+static struct net_buf *hid_buf_alloc(struct hid_device_data *const ddata,
+				     const uint8_t ep)
+{
+	struct net_buf *buf = NULL;
+	struct udc_buf_info *bi;
+
+	buf = net_buf_alloc(ddata->pool_out, K_NO_WAIT);
+	if (!buf) {
+		return NULL;
+	}
+
+	bi = udc_get_buf_info(buf);
+	memset(bi, 0, sizeof(struct udc_buf_info));
+	bi->ep = ep;
+
+	return buf;
+}
+
+static void hid_dev_output_handler(struct k_work *work)
+{
+	struct hid_device_data *ddata = CONTAINER_OF(work,
+						     struct hid_device_data,
+						     output_work);
+	struct usbd_class_data *c_data = ddata->c_data;
+	struct net_buf *buf;
+
+	if (!atomic_test_bit(&ddata->state, HID_DEV_CLASS_ENABLED)) {
+		return;
+	}
+
+	buf = hid_buf_alloc(ddata, hid_get_out_ep(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return;
+	}
+
+	if (usbd_ep_enqueue(c_data, buf)) {
+		LOG_ERR("Failed to enqueue buffer");
```

Comment:
Should we free buffer on error here?

Reply:
:fire: I will look into what happened here.

Reply:
It is already fixed. Thanks